### ISSUE TITLE
segmentation.MeanIoU is wrong

### ISSUE THREAD

**Issue by nrudakov** [Lightning-AI/torchmetrics#2558]
## 🐛 Bug
MeanIoU scored 56(!) over validation dataset
![image](https://github.com/Lightning-AI/torchmetrics/assets/5420651/a98e66c8-48fa-4e8d-a02a-ee09d1783be6)

Let's look at the source code:
https://github.com/Lightning-AI/torchmetrics/blob/596ed09b18c3d6786cb094c87da97838228461f3/src/torchmetrics/segmentation/mean_iou.py#L109

```py
def update(self, preds: Tensor, target: Tensor) -> None:
    """Update the state with the new data."""
    intersection, union = _mean_iou_update(preds, target, self.num_classes, self.include_background)
    score = _mean_iou_compute(intersection, union, per_class=self.per_class)
    self.score += score.mean(0) if self.per_class else score.mean()

def compute(self) -> Tensor:
    """Update the state with the new data."""
    return self.score  # / self.num_batches
```

There are several issues there:
- `self.score` is accumulated with each call of `update` method.
- `compute` method just returns the accumulated score
- description of `compute ` method is copypasted from `update` method
- division by `self.num_batches` is commented
- `num_batches` is definded on class level and not used anywhere else

Obviously, that code was neither reviewed nor tested, but somehow was released.

### To Reproduce

Call `metric.update(y_hat, y)` in `validation_step`
Log the metric in `on_validation_epoch_end`

### Expected behavior

MeanIoU computes correct value in [0, 1] range.

### Environment

- TorchMetrics 1.4.0.post0 from pip
- Python 3.11.9, PyTorch 2.3.0+cu121
- pytorch-lightning 2.2.4
- Windows 11

**Comment by DimitrisMantas** [comment]
I think now with zero_division having been added to JaccardIndex, there's no real need for MeanIoU at all.

The original motivation for this was that JaccardIndex used to assign a score of 0 to absent and ignored classes so you couldn't do classwise and macro averaging correctly. Now, you can just set zero_division to NaN and average to None, and get correct class scores. From there, you could do a nanmean to get the correct macro average.

**Comment by juliendenize** [comment]
I also noticed several issues using meaniou, typically if a class was not present in target and pred then the iou was not a nan but zero, diminuishing the average wrongly.

If interested I developed a meaniou metric based on torchmetrics API that I hope works correctly, I did not test it deeply and developed it fast, so it might be a bit ugly, but I checked that it matched HF results on few batches.

```python
from typing import Any

import torch
from torch import Tensor
from torchmetrics import Metric
from torchmetrics.functional.segmentation.mean_iou import _mean_iou_validate_args
from torchmetrics.utilities.checks import _check_same_shape


def _intersection_union_per_class(
    preds: Tensor,
    target: Tensor,
    num_classes: int,
    include_background: bool = False,
    index_tensors: bool = False,
) -> tuple[Tensor, Tensor]:
    """Computes the Intersection over Union (IoU) for each class."""
    _check_same_shape(preds, target)

    if index_tensors:
        preds = torch.nn.functional.one_hot(preds, num_classes=num_classes)
        target = torch.nn.functional.one_hot(target, num_classes=num_classes)

    if not include_background:
        preds = preds[..., 1:]
        target = target[..., 1:]

    preds = preds.bool()
    target = target.bool()

    reduce_axis = list(range(1, preds.ndim - 1))
    intersection = torch.sum(torch.logical_and(preds, target), dim=reduce_axis)
    target_sum = torch.sum(target, dim=reduce_axis)
    pred_sum = torch.sum(preds, dim=reduce_axis)
    union = target_sum + pred_sum - intersection
    union = torch.where(
        torch.logical_and(torch.logical_not(pred_sum), torch.logical_not(target_sum)), torch.nan, union
    )

    return intersection, union


class MeanIoU(Metric):
    intersection: Tensor
    union: Tensor
    full_state_update: bool = False  # type: ignore[override]
    is_differentiable: bool = False  # type: ignore[override]
    higher_is_better: bool = True  # type: ignore[override]
    plot_lower_bound: float = 0.0  # type: ignore[override]
    plot_upper_bound: float = 1.0  # type: ignore[override]

    def __init__(
        self,
        num_classes: int,
        include_background: bool = True,
        per_class: bool = False,
        index_tensors: bool = False,
        **kwargs: Any,
    ) -> None:
        Metric.__init__(self, **kwargs)
        _mean_iou_validate_args(num_classes, include_background, per_class)
        self.num_classes = num_classes
        self.include_background = include_background
        self.per_class = per_class
        self.index_tensors = index_tensors

        num_classes = num_classes - 1 if not include_background else num_classes
        self.add_state("intersection", default=torch.zeros(num_classes), dist_reduce_fx="sum")
        self.add_state("union", default=torch.zeros(num_classes), dist_reduce_fx="sum")

    def update(self, preds: Tensor, target: Tensor) -> None:
        intersection, union = _intersection_union_per_class(
            preds, target, self.num_classes, self.include_background, self.index_tensors
        )
        non_nan_union = torch.logical_not(torch.isnan(union))
        zero = torch.tensor(0, device=intersection.device)
        non_nan_class = torch.sum(non_nan_union, dim=0).bool()
        self.intersection += intersection.sum(0)
        self.union += torch.where(non_nan_class, torch.where(non_nan_union, union, zero).sum(0), zero)

    def compute(self) -> Tensor:
        zero = torch.tensor(0, device=self.device)
        iou = torch.where(
            torch.logical_not(torch.isnan(self.union)),
            torch.divide(self.intersection, self.union),
            torch.nan,
        )

        nan_score = torch.isnan(iou)
        iou_valid = torch.where(nan_score, torch.zeros_like(iou), iou)
        num_valid = torch.sum(torch.logical_not(nan_score), dim=0)
        any_valid = torch.ne(num_valid, torch.tensor(0, device=num_valid.device))
        sum_score = torch.where(
            any_valid,
            torch.sum(iou_valid, dim=0),
            torch.zeros(any_valid.shape, device=iou.device, dtype=iou_valid.dtype),
        )
        num_samples = torch.where(any_valid, num_valid, torch.zeros_like(num_valid))
        if self.per_class:
            return torch.where(torch.ne(num_samples, zero), torch.divide(sum_score, num_samples), torch.nan)
        else:
            valid_class = torch.ne(num_samples, zero)
            return torch.mean(torch.divide(sum_score[valid_class], num_samples[valid_class]))
```

I might do a PR if it makes sense for the owners

**Comment by juliendenize** [comment]
Hi, I also noticed that the MeanIoU was wrong during my experiments.

I developed something that seems to work on my side, which returns the same results as [evaluate's mean iou](https://huggingface.co/spaces/evaluate-metric/mean_iou) however based on @DimitrisMantas I wonder if it is relevant to submit a PR. I am not familiar enough with the JaccardIndex implementation in torchmetrics.

For reference, here is the undocumented code I developed (which has not been rigorusly tested for now), let me know if submitting a PR is something interesting for you, I'd gladly contribute to this repo.

```python
from typing import Any, Literal

import torch
from torch import Tensor
from torchmetrics import Metric

def _compute_intersection_and_union(
    preds: Tensor,
    target: Tensor,
    num_classes: int,
    include_background: bool = False,
    input_format: Literal["one-hot", "index", "predictions"] = "index",
) -> tuple[Tensor, Tensor]:
    if input_format in ["index", "predictions"]:
        if input_format == "predictions":
            preds = preds.argmax(1)
        preds = torch.nn.functional.one_hot(preds, num_classes=num_classes)
        target = torch.nn.functional.one_hot(target, num_classes=num_classes)

    if not include_background:
        preds[..., 0] = 0
        target[..., 0] = 0

    reduce_axis = list(range(1, preds.ndim - 1))
    intersection = torch.sum(torch.logical_and(preds, target), dim=reduce_axis)
    target_sum = torch.sum(target, dim=reduce_axis)
    pred_sum = torch.sum(preds, dim=reduce_axis)
    union = target_sum + pred_sum - intersection

    return intersection, union


class MeanIoU(Metric):
    def __init__(
        self,
        num_classes: int,
        include_background: bool = True,
        per_class: bool = False,
        input_format: Literal["one-hot", "index", "predictions"] = "index",
        **kwargs: Any,
    ) -> None:
        Metric.__init__(self, **kwargs)

        self.num_classes = num_classes
        self.include_background = include_background
        self.per_class = per_class
        self.input_format = input_format

        self.add_state("intersection", default=torch.zeros(num_classes), dist_reduce_fx="sum")
        self.add_state("union", default=torch.zeros(num_classes), dist_reduce_fx="sum")

    def update(self, preds: Tensor, target: Tensor) -> None:
        intersection, union = _compute_intersection_and_union(
            preds, target, self.num_classes, self.include_background, self.input_format
        )
        self.intersection += intersection.sum(0)
        self.union += union.sum(0)

    def compute(self) -> Tensor:
        iou_valid = torch.gt(self.union, 0)

        iou = torch.where(
            iou_valid,
            torch.divide(self.intersection, self.union),
            torch.nan,
        )

        if self.per_class:
            return iou
        else:
            return torch.mean(iou[iou_valid])
```

**Comment by vkinakh** [comment]
Hi, I have noticed that MeanIoU is incorrect, when updating it via calling `update` method, if I update it via `forward` method it works correct. I have looked into it, and it is because with default MeanIoU parameters `forward` method calls `_reduce_states` method, which updates `score` using the following formula:

`reduced = ((self._update_count - 1) * global_state + local_state).float() / self._update_count`,

where `global_state` is the score accumulated over previous steps and `local_state` is the score on current batch.

Here is the code to reproduce results

```python
import torch
from torchmetrics.segmentation import MeanIoU


def run():
    bs = 16
    num_classes = 3
    h = w = 128

    # one-hot, per_class=False
    miou_update = MeanIoU(num_classes=num_classes, input_format="one-hot", per_class=False)
    miou_call = MeanIoU(num_classes=num_classes, input_format="one-hot", per_class=False)

    # test 1, all ones
    img_pred = torch.ones(bs, num_classes, h, w, dtype=torch.long)
    img_target = torch.ones(bs, num_classes, h, w, dtype=torch.long)

    miou_update.update(img_pred, img_target)
    miou_call(img_pred, img_target)
    print(f"Update: {miou_update.compute()}, {miou_update.score}, {miou_update.update_count}")
    print(f"Call: {miou_call.compute()}, {miou_call.score}, {miou_call.update_count}")

    # test 2, square in the middle, 100% overlap
    img_pred = torch.zeros(bs, num_classes, h, w, dtype=torch.long)
    img_target = torch.zeros(bs, num_classes, h, w, dtype=torch.long)

    img_pred[:, :, 50:80, 50:80] = 1
    img_target[:, :, 50:80, 50:80] = 1

    miou_update.update(img_pred, img_target)
    miou_call(img_pred, img_target)
    print(f"Update: {miou_update.compute()}, {miou_update.score}, {miou_update.update_count}")
    print(f"Call: {miou_call.compute()}, {miou_call.score}, {miou_call.update_count}")

    # test 3, square in the middle, 50% overlap
    img_pred = torch.zeros(bs, num_classes, h, w, dtype=torch.long)
    img_target = torch.zeros(bs, num_classes, h, w, dtype=torch.long)

    img_pred[:, :, 50:80, 50:80] = 1
    img_target[:, :, 65:80, 50:80] = 1

    miou_update.update(img_pred, img_target)
    miou_call(img_pred, img_target)
    print(f"Update: {miou_update.compute()}, {miou_update.score}, {miou_update.update_count}")
    print(f"Call: {miou_call.compute()}, {miou_call.score}, {miou_call.update_count}")

    # test 4, square in the middle, 0% overlap
    img_pred = torch.zeros(bs, num_classes, h, w, dtype=torch.long)
    img_target = torch.zeros(bs, num_classes, h, w, dtype=torch.long)

    img_pred[:, :, 50:80, 50:80] = 1
    img_target[:, :, 80:100, 80:100] = 1

    miou_update.update(img_pred, img_target)
    miou_call(img_pred, img_target)
    print(f"Update: {miou_update.compute()}, { miou_update.score}, { miou_update.update_count}")
    print(f"Call: {miou_call.compute()}, { miou_call.score}, { miou_call.update_count}")

    # one-hot, per_class=True
    miou_update = MeanIoU(num_classes=num_classes, input_format="one-hot", per_class=True)
    miou_call = MeanIoU(num_classes=num_classes, input_format="one-hot", per_class=True)

    # test 5, all ones
    img_pred = torch.ones(bs, num_classes, h, w, dtype=torch.long)
    img_target = torch.ones(bs, num_classes, h, w, dtype=torch.long)

    miou_update.update(img_pred, img_target)
    miou_call(img_pred, img_target)
    print(f"Update: {miou_update.compute()}, {miou_update.score}, {miou_update.update_count}")
    print(f"Call: {miou_call.compute()}, {miou_call.score}, {miou_call.update_count}")

    # test 6, square in the middle, 100% overlap
    img_pred = torch.zeros(bs, num_classes, h, w, dtype=torch.long)
    img_target = torch.zeros(bs, num_classes, h, w, dtype=torch.long)

    img_pred[:, :, 50:80, 50:80] = 1
    img_target[:, :, 50:80, 50:80] = 1

    miou_update.update(img_pred, img_target)
    miou_call(img_pred, img_target)
    print(f"Update: {miou_update.compute()}, {miou_update.score}, {miou_update.update_count}")
    print(f"Call: {miou_call.compute()}, {miou_call.score}, {miou_call.update_count}")

    # test 7, square in the middle, 50% overlap
    img_pred = torch.zeros(bs, num_classes, h, w, dtype=torch.long)
    img_target = torch.zeros(bs, num_classes, h, w, dtype=torch.long)

    img_pred[:, :, 50:80, 50:80] = 1
    img_target[:, :, 65:80, 50:80] = 1

    miou_update.update(img_pred, img_target)
    miou_call(img_pred, img_target)
    print(f"Update: {miou_update.compute()}, {miou_update.score}, {miou_update.update_count}")
    print(f"Call: {miou_call.compute()}, {miou_call.score}, {miou_call.update_count}")

    # test 8, square in the middle, 0% overlap
    img_pred = torch.zeros(bs, num_classes, h, w, dtype=torch.long)
    img_target = torch.zeros(bs, num_classes, h, w, dtype=torch.long)

    img_pred[:, :, 50:80, 50:80] = 1
    img_target[:, :, 80:100, 80:100] = 1

    miou_update.update(img_pred, img_target)
    miou_call(img_pred, img_target)
    print(f"Update: {miou_update.compute()}, {miou_update.score}, {miou_update.update_count}")
    print(f"Call: {miou_call.compute()}, {miou_call.score}, {miou_call.update_count}")

    # index, per_class=False
    miou_update = MeanIoU(num_classes=num_classes, input_format="index", per_class=True)
    miou_call = MeanIoU(num_classes=num_classes, input_format="index", per_class=True)

    # test 9, all ones
    img_pred = torch.ones(bs, h, w, dtype=torch.long)
    img_target = torch.ones(bs, h, w, dtype=torch.long)

    miou_update.update(img_pred, img_target)
    miou_call(img_pred, img_target)
    print(f"Update: {miou_update.compute()}, {miou_update.score}, {miou_update.update_count}")
    print(f"Call: {miou_call.compute()}, {miou_call.score}, {miou_call.update_count}")

    # test 10, square in the middle, 100% overlap
    img_pred = torch.zeros(bs, h, w, dtype=torch.long)
    img_target = torch.zeros(bs, h, w, dtype=torch.long)

    img_pred[:, 50:80, 50:80] = 1
    img_target[:, 50:80, 50:80] = 1

    miou_update.update(img_pred, img_target)
    miou_call(img_pred, img_target)
    print(f"Update: {miou_update.compute()}, {miou_update.score}, {miou_update.update_count}")
    print(f"Call: {miou_call.compute()}, {miou_call.score}, {miou_call.update_count}")

    # test 11, square in the middle, 50% overlap
    img_pred = torch.zeros(bs, h, w, dtype=torch.long)
    img_target = torch.zeros(bs, h, w, dtype=torch.long)

    img_pred[:, 50:80, 50:80] = 1
    img_target[:, 65:80, 50:80] = 1

    miou_update.update(img_pred, img_target)
    miou_call(img_pred, img_target)
    print(f"Update: {miou_update.compute()}, {miou_update.score}, {miou_update.update_count}")
    print(f"Call: {miou_call.compute()}, {miou_call.score}, {miou_call.update_count}")

    # test 12, square in the middle, 0% overlap
    img_pred = torch.zeros(bs, h, w, dtype=torch.long)

    img_pred[:, 50:80, 50:80] = 1
    img_target[:, 80:100, 80:100] = 1

    miou_update.update(img_pred, img_target)
    miou_call(img_pred, img_target)
    print(f"Update: {miou_update.compute()}, {miou_update.score}, {miou_update.update_count}")
    print(f"Call: {miou_call.compute()}, {miou_call.score}, {miou_call.update_count}")
```

**Comment by rittik9** [comment]
Hi, I'm new to open-source contributions, and I want to start by working on this issue. @Borda could you please assign it to me?
